drivers/led/lp5562.c

```diff
@@ -0,0 +1,837 @@
+	LP5562_ENGINE_MODE_STEP = 0x01,
+	LP5562_ENGINE_MODE_RUN = 0x02,
+	LP5562_ENGINE_MODE_EXEC = 0x03,
+};
+
+/* Fading directions for programs executed by the engines. */
+enum lp5562_engine_fade_dirs {
+	LP5562_FADE_UP = 0x00,
+	LP5562_FADE_DOWN = 0x01,
+};
+
+struct lp5562_data {
+	struct device *i2c;
+	struct led_data dev_data;
+};
+
+/*
+ * @brief Get the register for the given LED channel used to directly write a
+ *	brightness value instead of using the execution engines.
+ *
+ * @param channel LED channel.
+ * @param source Buffer for the register.
+ *
+ * @return 0 if successful, -ERRNO otherwise.
+ */
+static int lp5562_get_pwm_reg(enum lp5562_led_channels channel, u8_t *reg)
+{
+	switch (channel) {
+	case LP5562_CHANNEL_W:
+		*reg = LP5562_W_PWM;
+		break;
+	case LP5562_CHANNEL_R:
+		*reg = LP5562_R_PWM;
+		break;
+	case LP5562_CHANNEL_G:
+		*reg = LP5562_G_PWM;
+		break;
+	case LP5562_CHANNEL_B:
+		*reg = LP5562_B_PWM;
+		break;
+	default:
+		SYS_LOG_ERR("Invalid channel given.");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+/*
+ * @brief Get the base address for programs of the given execution engine.
+ *
+ * @param engine Engine the base address is requested for.
+ * @param base_addr Buffer for the base address.
+ *
+ * @return 0 if successful, -ERRNO otherwise.
+ */
+static int lp5562_get_engine_ram_base_addr(enum lp5562_led_sources engine,
+					   u8_t *base_addr)
+{
+	switch (engine) {
+	case LP5562_SOURCE_ENGINE_1:
+		*base_addr = LP5562_PROG_MEM_ENG1_BASE;
+		break;
+	case LP5562_SOURCE_ENGINE_2:
+		*base_addr = LP5562_PROG_MEM_ENG2_BASE;
+		break;
+	case LP5562_SOURCE_ENGINE_3:
+		*base_addr = LP5562_PROG_MEM_ENG3_BASE;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+/*
+ * @brief Helper to get the register bit shift for the execution engines.
+ *
+ * The engine with the highest index is placed on the lowest two bits in the
+ * OP_MODE and ENABLE registers.
+ *
+ * @param engine Engine the shift is requested for.
+ * @param shift Buffer for the shift value.
+ *
+ * @param 0 if successful, -ERRNO otherwise.
+ */
+static int lp5562_get_engine_reg_shift(enum lp5562_led_sources engine,
+				       u8_t *shift)
+{
+	switch (engine) {
+	case LP5562_SOURCE_ENGINE_1:
+		*shift = 4;
+		break;
+	case LP5562_SOURCE_ENGINE_2:
+		*shift = 2;
+		break;
+	case LP5562_SOURCE_ENGINE_3:
+		*shift = 0;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+/*
+ * @brief Convert a time in milliseconds to a combination of prescale and
+ *	step_time for the execution engine programs.
+ *
+ * @param data Capabilities of the driver.
+ * @param ms Time to be converted in milliseconds.
+ * @param prescale Buffer for the prescaled value.
+ * @param step_time Buffer for the step_time value.
+ *
+ * @param 0 if successful, -ERRNO otherwise.
+ */
+static int lp5562_ms_to_prescale_and_step(struct led_data *data, u32_t ms,
+					  u8_t *prescale, u8_t *step_time)
+{
+	if ((ms < data->min_period) || (ms > data->max_period)) {
```

Comment:
Check here is superfluous

Reply:
hm, I am no fan of allowing invalid arguments, but I can remove the check and add a comment.

Reply:
What I meant was, the argument check is already done in the caller `lp5562_program_ramp`, so you don't need to do the same in this function also. That's why I mentioned it as superfluous!